src/nvim/eval.c

```diff
@@ -6007,6 +6009,27 @@ static void rettv_dict_alloc(typval_T *rettv)
   ++d->dv_refcount;
 }
 
+/// Clear the all the keys of a Dictionary.
+/// The Dictionary remains a valid empty Dictionary.
+void dict_clear(dict_T *d) {
+  hash_lock(&d->dv_hashtab);
+  assert(d->dv_hashtab.ht_locked > 0);
+
+  size_t todo = d->dv_hashtab.ht_used;
+  for (hashitem_T *hi = d->dv_hashtab.ht_array; todo > 0; ++hi) {
+    if (!HASHITEM_EMPTY(hi)) {
+      // Remove the item before deleting it, just in case there is
+      // something recursive causing trouble.
+      dictitem_T *di = HI2DI(hi);
+      hash_remove(&d->dv_hashtab, hi);
+      dictitem_free(di);
```

Comment:
`di` is used only once, here. It would be logical to remove this variable and write `HI2DI(hi)` in this line.